### PR TITLE
Add matrix routines for multi-qubit gates.

### PR DESCRIPTION
**Pull request by sergeisakov**

### CODE REVIEW

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Does this affect performance at all for the usual execution of qsim (i.e. with gates having 2 qubits at most)?

File: lib/bits.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+}  // namespace bits
+}  // namespace qsim
+
+#else  // __BMI2__
+
+namespace qsim {
+namespace bits {
+
+template <typename I>
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply lib/bits.h]: This is just an integer. I replaced `I` with `Integer`.

File: lib/gate_appl.h [hunk 2]
```diff
@@ -23,45 +23,38 @@
...
-template <typename Gate, typename Array2>
-inline void CalcMatrix2(const std::vector<Gate*>& gates, Array2& matrix) {
-  Matrix2SetId(matrix);
+namespace detail {
 
-  for (auto pgate : gates) {
-    Matrix2Multiply(pgate->matrix, matrix);
-  }
-}
+template <typename fp_type, typename Gate>
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply lib/gate_appl.h]: The idea was that this function is not a part of public API. However, it is probably better if this function is a part of public API. I added a comment and moved this function to fuser.h.

**Comment by sergeisakov** [comment]
> Does this affect performance at all for the usual execution of qsim (i.e. with gates having 2 qubits at most)?

This should not affect perfomance, especially for large enough circuits (say, for more than 8 qubits).

**Review by 95-martin-orion — APPROVED** [review]

**Pull request merged**